test/unit/org/apache/cassandra/metrics/CompactionMetricsTest.java

```diff
@@ -0,0 +1,249 @@
+            assertTrue(compactionManager.getMetrics().pendingTasks.getValue() > initialPendingTasks);
+        }
+        finally
+        {
+            tct.abort();
+        }
+    }
+
+    @Test
+    public void testCompactionMetricsForCompactionsAborted() throws Throwable
+    {
+        createTable("create table %s (id int, id2 int, t text, primary key (id, id2))");
+        execute("insert into %s (id, id2) values (1, 1)");
+        flush();
+
+        for(File dir : getCurrentColumnFamilyStore().getDirectories().getCFDirectories()){
+            DisallowedDirectories.maybeMarkUnwritable(dir);
+        }
+
+        try
+        {
+            getCurrentColumnFamilyStore().forceMajorCompaction();
+        }
+        catch (Exception e){
+            logger.info("Exception will be thrown when compacting with unwritable directories.");
+        }
+        assertTrue(compactionManager.getMetrics().compactionsAborted.getCount() > 0);
+    }
+
+    @Test
+    public void testCompactionMetricsForDroppedSSTables () throws Throwable
+    {
+        // assertTrue(compactionManager.getMetrics().compactionsReduced.getCount() > 0);
+        // assertTrue(compactionManager.getMetrics().sstablesDropppedFromCompactions.getCount() > 0);
+    }
+
+    private List<SSTableReader> createSSTables(ColumnFamilyStore cfs, int count, int startGeneration)
+    {
+        List<SSTableReader> sstables = new ArrayList<>();
+        for (int i = 0; i < count; i++)
+        {
+            long first = i * 10;
+            long last = (i + 1) * 10 - 1;
+            sstables.add(MockSchema.sstable(startGeneration + i, 0, true, first, last, cfs));
+        }
+        cfs.disableAutoCompaction();
+        cfs.addSSTables(sstables);
+        return sstables;
+    }
+
+    private static class TestCompactionTask
+    {
+        private final Set<SSTableReader> sstables;
+        ActiveCompactionsTracker activeCompactions = compactionManager.active;
+        OperationType compactionType = OperationType.COMPACTION;
+        private final ColumnFamilyStore cfs;
+        private LifecycleTransaction txn;
+        private CompactionController controller;
+        private CompactionIterator ci;
+        private List<ISSTableScanner> scanners;
+
+        public TestCompactionTask(ColumnFamilyStore cfs, Set<SSTableReader> sstables)
+        {
+            this.cfs = cfs;
+            this.sstables = sstables;
+        }
+
+        public void start()
+        {
+            scanners = sstables.stream().map(SSTableReader::getScanner).collect(Collectors.toList());
+            txn = cfs.getTracker().tryModify(sstables, OperationType.COMPACTION);
+            assertNotNull(txn);
+            controller = new CompactionController(cfs, sstables, Integer.MIN_VALUE);
+            ci = new CompactionIterator(txn.opType(), scanners, controller, FBUtilities.nowInSeconds(), UUID.randomUUID());
+            activeCompactions.beginCompaction(ci);
```

Comment:
I feel like we're doing a lot of mocking here just to put a CompactionInfo.Holder in a set. I think all this could be replaced with a call to `beginCompaction` with an anonymous Holder as is done in some other tests:
https://github.com/apache/cassandra/blob/46ee939b957528185dc6bbd3028c1d6e695163e7/test/unit/org/apache/cassandra/db/repair/PendingAntiCompactionTest.java#L603-L614

However, this still only goes so far as to test the gauges, but not the actual production mechanisms that tick them in a running server. I think we would be best off concocting tests that use actual compactions to test their integration with metrics. Much like you did for `CompletedTasks`. Pending metrics could be a bit more challenging due to timing, but we should be able to create the right conditions through data size, compaction throttling, etc.